### DISCUSSION TITLE
Could `futures::executor::block_on()` lead to deadlock when being invoked in a Monoio task?

### DISCUSSION THREAD

**Discussion by SteveLauC** [bytedance/monoio#312] [Q&A] [answered]
Hi, I would like to know if a use of `futures::executor::block_on()` within a Monoio task could potentially lead to a deadlock? If so, would you like to explain the causes?

BTW, [Tokio's answer is yes](https://github.com/tokio-rs/tokio/issues/2376)

**Comment by ethe** [comment, marked as answer]
The same is true in monoio. Running the io-bound future in `futures::block_on` will cause the monoio reactor to lose the opportunity to react the io event, and `futures::block_on` will never be ready.

**Reply by SteveLauC** [reply]
This only happens if the task executed by `futures::executor::block_on()` **depends on** a task that is executed in your Monoio runtime though.

**Reply by ethe** [reply]
That's another case, io-bound or awaiting another monoio task will lead to deadlock.

**Reply by SteveLauC** [reply]
Oops, sorry that I didn't notice it, do you mean something like:

```rs
#[monoio::main]
async fn main() {
    let file = monoio::fs::File::open("Cargo.toml").await.unwrap();
    futures::executor::block_on(async move {
        let (result, buf) = file.read_at(vec![0_u8; 10], 0).await;
        println!("result {:?}, buf {:?}", result, buf);
    });
}
```

The above code snippet indeed blocks in my test. 🤔
